spyder/widgets/pathmanager.py

```diff
@@ -21,285 +23,431 @@
+        self.listwidget.setCurrentRow(0)
+        self.original_path_dict = self.get_path_dict()
+        self.refresh()
 
     @Slot()
     def synchronize(self):
         """
         Synchronize Spyder's path list with PYTHONPATH environment variable
-        Only apply to: current user, on Windows platforms
+        Only apply to: current user, on Windows platforms.
         """
-        answer = QMessageBox.question(self, _("Synchronize"),
+        answer = QMessageBox.question(
+            self,
+            _("Synchronize"),
             _("This will synchronize Spyder's path list with "
-                    "<b>PYTHONPATH</b> environment variable for current user, "
-                    "allowing you to run your Python modules outside Spyder "
-                    "without having to configure sys.path. "
-                    "<br>Do you want to clear contents of PYTHONPATH before "
-                    "adding Spyder's path list?"),
+              "<b>PYTHONPATH</b> environment variable for current user, "
+              "allowing you to run your Python modules outside Spyder "
+              "without having to configure sys.path. "
+              "<br>"
+              "Do you want to clear contents of PYTHONPATH before "
+              "adding Spyder's path list?"),
             QMessageBox.Yes | QMessageBox.No | QMessageBox.Cancel)
+
         if answer == QMessageBox.Cancel:
             return
         elif answer == QMessageBox.Yes:
             remove = True
         else:
             remove = False
-        from spyder.utils.environ import (get_user_env, set_user_env,
-                                          listdict2envdict)
+
+        from spyder.utils.environ import (get_user_env, listdict2envdict,
+                                          set_user_env)
         env = get_user_env()
+
+        # Includes read only paths
+        active_path = tuple(k for k, v in self.get_path_dict().items() if v)
+
         if remove:
-            ppath = self.active_pathlist+self.ro_pathlist
+            ppath = active_path
         else:
             ppath = env.get('PYTHONPATH', [])
+
             if not isinstance(ppath, list):
                 ppath = [ppath]
-            ppath = [path for path in ppath
-                     if path not in (self.active_pathlist+self.ro_pathlist)]
-            ppath.extend(self.active_pathlist+self.ro_pathlist)
+
+            ppath = tuple(p for p in ppath if p not in active_path)
+            ppath = ppath + active_path
+
         env['PYTHONPATH'] = ppath
         set_user_env(listdict2envdict(env), parent=self)
 
-    def get_path_list(self):
-        """Return path list (does not include the read-only path list)"""
-        return self.pathlist
+    def get_path_dict(self, read_only=False):
+        """
+        Return an ordered dict with the path entries as keys and the active
+        state as the value.
+
+        If `read_only` is True, the read_only entries are also included.

```

Comment:
You could clarify here that `read_only` refers to the current project path. Else it's hard to understand what that's used for.